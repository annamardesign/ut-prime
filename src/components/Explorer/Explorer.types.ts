import {createUseStyles} from 'react-jss';
import React from 'react';
import { DataTableProps } from '../prime';

import type {Schema, Dropdowns, Action, WidgetReference} from '../types';

export interface Props {
    keyField?: string;
    resultSet?: string;
    schema: Schema;
    columns: WidgetReference[];
    fetch: (params: {
        [resultSet: string]: {},
        orderBy: {
            field: string,
            dir: string
        }[],
        paging: {
            pageSize: number,
            pageNumber: number
        }
    }) => Promise<{
        pagination?: {
            recordsTotal: number
        },
        [data: string]: any
    }>;
    subscribe?: (callback: (rows: any) => void) => () => void;
    onDropdown?: (params: string[]) => Promise<Dropdowns>,
    className?: string;
    details?: {};
    actions?: Action[];
    filter?: {};
    index?: {};
    showFilter?: boolean;
    pageSize?: number;
    table?: DataTableProps;
}

export const useStyles = createUseStyles({
    explorer: {
        '& .p-datatable-wrapper': {
            overflowX: 'auto'
        }
    },
    details: {
        marginBottom: 15
    },
    detailsLabel: {},
    detailsValue: {}
});

export type ComponentProps = React.FC<Props>
