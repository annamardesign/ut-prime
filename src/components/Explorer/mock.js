const filteredItems = (filterBy) => {
    return Object.entries(filterBy).reduce((items, [key, value]) => {
        return items.filter(i => {
            return i[key] === value || (typeof value === 'string' && i[key].toString().toLowerCase().includes(value.toLowerCase()));
        });
    }, [...Array(55).keys()].map(number => ({
        id: number,
        name: `Item ${number}`,
        size: number * 10
    })));
};

const compare = ({field, dir, smaller = {ASC: -1, DESC: 1}[dir]}) => function compare(a, b) {
    if (a[field] < b[field]) return smaller;
    if (a[field] > b[field]) return -smaller;
    return 0;
};

export const fetchItems = filters => {
    const items = filteredItems(filters.items).sort(compare(filters.orderBy));
    return Promise.resolve({
        items: items.slice(filters.paging.pageNumber * filters.paging.pageSize, (filters.paging.pageNumber + 1) * filters.paging.pageSize),
        pagination: {
            recordsTotal: items.length
        }
    });
};
