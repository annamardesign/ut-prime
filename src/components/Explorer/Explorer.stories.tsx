import React from 'react';
import { withReadme } from 'storybook-readme';

import Wrap from '../test/wrap';

// @ts-ignore: md file and not a module
import README from './README.md';
import Explorer from './index';

export default {
    title: 'Explorer',
    component: Explorer,
    decorators: [withReadme(README)]
};

const state = {
};

export const Basic: React.FC<{}> = () => <Wrap state={state}>
    <div style={{height: 500}}>
        <Explorer
            fetch={() => Promise.resolve({
                items: [...Array(50).keys()].map(number => ({
                    id: number,
                    name: `Item ${number}`,
                    size: number * 10
                }))
            })}
            keyField='id'
            resultSet='items'
            fields={[{
                dataField: 'name',
                caption: 'Name'
            }, {
                dataField: 'size',
                caption: 'Size'
            }]}
            details={{
                name: 'Name'
            }}
        >
            <div>Navigation component</div>
        </Explorer>
    </div>
</Wrap>;
