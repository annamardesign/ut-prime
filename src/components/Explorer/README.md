# Explorer

## How to use

```jsx
import Explorer from 'ut-front-devextreme/core/Explorer';

<Explorer className='name' />;
```

## Props

- **className** - class applied to root `Explorer` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| className | string   |              | no         |
