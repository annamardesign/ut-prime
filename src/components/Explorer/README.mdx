# Explorer

import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

Shows a collection of items in a tabular format,
allowing filtering, and performing actions on one or multiple items.
Item details can be shown in the side pane.

It is named after the Windows File Explorer, with which it
aims to have similarities.

## How to use

```jsx
import Explorer from 'ut-prime/core/Explorer';

<Explorer
    fetch={fetch}
    keyField='id'
    resultSet='items'
    properties={properties}
    columns={columns}
    details={details}
/>
```

<Canvas>
    <Story id="explorer--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
