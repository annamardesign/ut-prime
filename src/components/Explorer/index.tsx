import React from 'react';
import DataGrid, {Selection} from 'devextreme-react/data-grid';
import Box, { Item } from 'devextreme-react/box';
import DataSource from 'devextreme/data/data_source';
import CustomStore from 'devextreme/data/custom_store';
import Toolbar from 'devextreme-react/toolbar';
import Drawer from 'devextreme-react/drawer';

import { Styled, StyledType } from './Explorer.types';
import useToggle from '../hooks/useToggle';

const Explorer: StyledType = ({
    classes,
    className,
    keyField,
    fetch,
    fields,
    resultSet,
    children,
    details
}) => {
    const organizations = React.useMemo(() => new DataSource({
        store: new CustomStore({
            key: keyField,
            load: async() => (await fetch({}))[resultSet]
        })
    }), [fetch]);
    const [current, setCurrent] = React.useState({});
    const handleRowFocused = React.useCallback(({row}) => setCurrent(row.data), [setCurrent]);
    const Details = () =>
        <div style={{ width: '200px' }}>{
            Object.entries(details).map(([name, value], index) =>
                <div className={classes.details} key={index}>
                    <div className={classes.detailsLabel}>{value}</div>
                    <div className={classes.detailsValue}>{current[name]}</div>
                </div>
            )
        }</div>;

    const [navigationOpened, navigationToggle] = useToggle(true);
    const [detailsOpened, detailsToggle] = useToggle(true);
    const navigation = React.useCallback(
        () => <div style={{ width: '200px' }}>{children}</div>,
        [children]
    );

    return (
        <Box direction='col' height='100%' className={className}>
            <Item baseSize={76}>
                <Toolbar
                    items={[{
                        widget: 'dxButton',
                        location: 'before',
                        options: {
                            icon: 'menu',
                            onClick: navigationToggle
                        }
                    }, {
                        widget: 'dxButton',
                        location: 'after',
                        options: {
                            icon: 'menu',
                            onClick: detailsToggle
                        }
                    }]}
                />
            </Item>
            <Item ratio={1}>
                <Drawer
                    opened={navigationOpened}
                    openedStateMode='shrink'
                    position='left'
                    revealMode='slide'
                    component={navigation}
                >
                    <Drawer
                        opened={detailsOpened}
                        openedStateMode='shrink'
                        position='right'
                        revealMode='slide'
                        component={Details}
                    >
                        <DataGrid
                            height='100%'
                            focusedRowEnabled
                            onFocusedRowChanged={handleRowFocused}
                            dataSource={organizations}
                            defaultColumns={fields}
                            showBorders={true}
                        >
                            <Selection
                                mode='multiple'
                                selectAllMode='page'
                                showCheckBoxesMode='onClick'
                            />
                        </DataGrid>
                    </Drawer>
                </Drawer>
            </Item>
        </Box>
    );
};

export default Styled(Explorer);
