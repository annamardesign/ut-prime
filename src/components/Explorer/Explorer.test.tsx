import React from 'react';
import {waitFor} from '@testing-library/react';

import { render } from '../test';
import Explorer from './index';
import {fetchItems} from './mock';

describe('<Explorer />', () => {
    it('render equals snapshot', async() => {
        const { getByTestId } = render(<Explorer
            fetch={fetchItems}
            keyField='id'
            resultSet='items'
            fields={[{
                field: 'name',
                title: 'Name',
                filter: true,
                sort: true
            }, {
                field: 'size',
                title: 'Size'
            }]}
            details={{
                name: 'Name'
            }}
        />);
        await waitFor(() => expect(getByTestId('ut-front-test').querySelector('tr.p-selectable-row')).toBeTruthy());
        expect(getByTestId('ut-front-test')).toMatchSnapshot();
    });
});
