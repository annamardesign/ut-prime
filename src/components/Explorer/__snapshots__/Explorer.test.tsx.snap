// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Explorer /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="dx-box-flex dx-box dx-widget dx-visibility-change-handler dx-collection"
    style="display: flex; flex-direction: column; height: 100%; justify-content: flex-start; align-items: stretch;"
  >
    <div
      class="dx-item dx-box-item"
      style="display: flex; max-height: none; min-height: 0px; flex-basis: 76px; flex-shrink: 1;"
    >
      <div
        class="dx-item-content dx-box-item-content"
        style="width: auto; height: auto; display: flex; flex-basis: 0px; flex-grow: 1; flex-direction: column;"
      >
        <div
          class="dx-toolbar dx-widget dx-visibility-change-handler dx-collection"
          role="toolbar"
        >
          <div
            class="dx-toolbar-items-container"
          >
            <div
              class="dx-toolbar-before"
            >
              <div
                class="dx-item dx-toolbar-item dx-toolbar-button"
              >
                <div
                  class="dx-item-content dx-toolbar-item-content"
                >
                  <div
                    aria-label="menu"
                    class="dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-icon"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="dx-button-content"
                    >
                      <i
                        class="dx-icon dx-icon-menu"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="dx-toolbar-center"
            />
            <div
              class="dx-toolbar-after"
            >
              <div
                class="dx-item dx-toolbar-item dx-toolbar-button"
              >
                <div
                  class="dx-item-content dx-toolbar-item-content"
                >
                  <div
                    aria-label="menu"
                    class="dx-button dx-button-normal dx-button-mode-contained dx-widget dx-button-has-icon"
                    role="button"
                    tabindex="0"
                  >
                    <div
                      class="dx-button-content"
                    >
                      <i
                        class="dx-icon dx-icon-menu"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="dx-item dx-box-item"
      style="display: flex; max-height: none; min-height: 0px; flex-basis: 0px; flex-grow: 1; flex-shrink: 1;"
    >
      <div
        class="dx-item-content dx-box-item-content"
        style="width: auto; height: auto; display: flex; flex-basis: 0px; flex-grow: 1; flex-direction: column;"
      >
        <div
          class="dx-drawer dx-widget dx-visibility-change-handler dx-drawer-opened dx-drawer-shrink dx-drawer-slide dx-drawer-left"
        >
          <div
            class="dx-drawer-wrapper"
          >
            <div
              class="dx-drawer-panel-content"
            />
            <div
              class="dx-drawer-content"
            >
              <div
                class="dx-drawer-shader dx-state-invisible"
              />
              <div
                class="dx-drawer dx-widget dx-visibility-change-handler dx-drawer-opened dx-drawer-shrink dx-drawer-slide dx-drawer-right"
              >
                <div
                  class="dx-drawer-wrapper"
                >
                  <div
                    class="dx-drawer-content"
                  >
                    <div
                      class="dx-drawer-shader dx-state-invisible"
                    />
                    <div
                      class="dx-widget dx-visibility-change-handler"
                      role="presentation"
                      style="height: 100%;"
                    >
                      <div
                        aria-label="Data grid"
                        class="dx-datagrid dx-gridbase-container dx-datagrid-borders"
                        role="grid"
                      >
                        <div
                          class="dx-hidden"
                        />
                        <div
                          class="dx-hidden"
                        />
                        <div
                          class="dx-hidden"
                        />
                        <div
                          class="dx-hidden"
                        />
                        <div
                          class="dx-datagrid-headers dx-datagrid-nowrap"
                          role="presentation"
                        >
                          <div
                            class="dx-datagrid-content dx-datagrid-scroll-container"
                            role="presentation"
                          >
                            <table
                              class="dx-datagrid-table dx-datagrid-table-fixed"
                              role="presentation"
                            >
                              <colgroup>
                                <col
                                  style="width: auto;"
                                />
                                <col />
                                <col />
                              </colgroup>
                              <tbody
                                role="presentation"
                              />
                            </table>
                          </div>
                        </div>
                        <div
                          class="dx-datagrid-rowsview dx-datagrid-nowrap dx-empty"
                          role="presentation"
                          tabindex="0"
                        >
                          <div
                            class="dx-datagrid-content"
                          >
                            <table
                              class="dx-datagrid-table dx-datagrid-table-fixed dx-select-checkboxes-hidden"
                              role="presentation"
                            >
                              <colgroup>
                                <col
                                  style="width: auto;"
                                />
                                <col />
                                <col />
                              </colgroup>
                              <tbody
                                role="presentation"
                              >
                                <tr
                                  class="dx-row dx-freespace-row dx-column-lines"
                                  role="presentation"
                                >
                                  <td
                                    class="dx-command-select"
                                    style="text-align: center;"
                                  />
                                  <td
                                    style="text-align: left;"
                                  />
                                  <td
                                    style="text-align: left;"
                                  />
                                </tr>
                              </tbody>
                            </table>
                          </div>
                        </div>
                        <div
                          class="dx-overlay dx-widget dx-state-invisible dx-visibility-change-handler dx-loadpanel"
                        >
                          <div
                            aria-hidden="true"
                            class="dx-overlay-content"
                            style="width: 200px; height: 90px;"
                          />
                        </div>
                        <div
                          class="dx-hidden"
                        />
                        <div />
                        <div
                          class="dx-hidden"
                        />
                        <div
                          class="dx-hidden"
                        />
                        <div
                          class="dx-datagrid-drag-header dx-datagrid-text-content dx-widget dx-state-invisible"
                        />
                        <div
                          class="dx-context-menu dx-has-context-menu dx-widget dx-visibility-change-handler dx-collection dx-datagrid"
                          role="presentation"
                        />
                        <div
                          class="dx-header-filter-menu"
                        />
                        <div />
                      </div>
                    </div>
                  </div>
                  <div
                    class="dx-drawer-panel-content"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
